prow/pod-utils/gcs/upload.go

```diff
@@ -83,11 +91,27 @@ func upload(dtw destToWriter, uploadTargets map[string]UploadFunc) error {
 		log := logrus.WithField("dest", dest)
 		log.Info("Queued for upload")
 		go func(f UploadFunc, writer dataWriter, log *logrus.Entry) {
-			sem.Acquire(context.Background(), 1)
-			defer sem.Release(1)
 			defer group.Done()
-			if err := f(writer); err != nil {
+
+			var err error
+
+			for retryIndex := 1; retryIndex <= retryCount; retryIndex++ {
+				sem.Acquire(context.Background(), 1)
```

Comment:
Don't change the locking code please, just acquiring a  lock in the beginning of the func and releasing it via `defer` is completely fine, not using `defer` for the lock release is prone to introduce bugs

Reply:
I had moved the semaphore inside the retry loop so that other files can progress while a failed one sleeps before reattempt.
Agree with you about the need to use defer instead. The defer gets triggered at function boundary and not at for loop level. I've created a small inline function to use defer properly. 
Please let me know what you think.